substrate-archive/src/actors/workers/extrinsics_decoder.rs

```diff
@@ -0,0 +1,144 @@
+
+use arc_swap::ArcSwap;
+use async_std::task;
+use itertools::Itertools;
+use sqlx::PgPool;
+use std::{collections::HashMap, sync::Arc};
+use xtra::prelude::*;
+
+use desub::Decoder;
+
+use crate::{
+	actors::{
+		workers::database::{DatabaseActor, GetState},
+		SystemConfig,
+	},
+	database::{models::ExtrinsicsModel, queries},
+	error::{ArchiveError, Result},
+	types::BatchExtrinsics,
+};
+
+pub struct ExtrinsicsDecoder {
+	pool: PgPool,
+	addr: Address<DatabaseActor>,
+	max_block_load: u32,
+	decoder: Arc<Decoder>,
+	/// Cache of blocks where runtime upgrades occurred.
+	/// number -> spec
+	upgrades: ArcSwap<HashMap<u32, u32>>,
+}
+
+impl ExtrinsicsDecoder {
+	pub async fn new<B: Send + Sync, Db: Send + Sync>(
+		config: &SystemConfig<B, Db>,
+		addr: Address<DatabaseActor>,
+	) -> Result<Self> {
+		let max_block_load = config.control.max_block_load;
+		let chain = config.persistent_config.chain();
+		let pool = addr.send(GetState::Pool).await??.pool();
+		let decoder = Arc::new(Decoder::new(chain));
+		let mut conn = pool.acquire().await?;
+		let upgrades = ArcSwap::from_pointee(queries::upgrade_blocks_from_spec(&mut conn, 0).await?);
+		log::info!("Started extrinsic decoder");
+		Ok(Self { pool, addr, max_block_load, decoder, upgrades })
+	}
+
+	async fn crawl_missing_extrinsics(&mut self) -> Result<()> {
+		let mut conn = self.pool.acquire().await?;
+		let blocks = queries::missing_extrinsic_blocks(&mut conn, self.max_block_load).await?;
+		log::info!("Indexing {} missing extrinsic blocks", blocks.len());
+
+		let versions: Vec<u32> =
+			blocks.iter().filter(|b| !self.decoder.has_version(&b.3)).map(|(_, _, _, v)| *v).unique().collect();
+		// above and below line are separate to let immutable ref to `self.decoder` to go out of scope.
+		for version in versions.iter() {
+			let metadata = queries::metadata(&mut conn, *version as i32).await?;
+			log::debug!("Registering version {}", version);
+			Arc::get_mut(&mut self.decoder)
+				.expect("Actors guarantee one reference; qed")
+				.register_version(*version, &metadata)?;
+			log::debug!("Registered version {}", version);
+		}
+
+		if self.upgrades.load().iter().max_by(|a, b| a.1.cmp(b.1)).map(|(_, v)| v)
+			< blocks.iter().map(|&(_, _, _, v)| v).max().as_ref()
+		{
+			self.update_upgrade_blocks().await?;
+		}
+		let decoder = self.decoder.clone();
+		let upgrades = self.upgrades.load().clone();
+		let extrinsics =
+			task::spawn_blocking(move || Ok::<_, ArchiveError>(Self::decode(&decoder, blocks, &upgrades))).await??;
+
+		self.addr.send(BatchExtrinsics::new(extrinsics)).await?;
+		Ok(())
+	}
+
+	fn decode(
+		decoder: &Decoder,
+		blocks: Vec<(u32, Vec<u8>, Vec<u8>, u32)>,
+		upgrades: &HashMap<u32, u32>,
+	) -> Result<Vec<ExtrinsicsModel>> {
+		let mut extrinsics = Vec::new();
+		for (number, hash, ext, spec) in blocks.into_iter() {
+			if let Some(version) = upgrades.get(&number) {
+				let previous = itertools::sorted(upgrades.values())
```

Comment:
Why is clippy so upset about `itertools` not being used?

Reply:
:shrug: 
i changed it to not use itertools::sorted but instead .sorted() from the trait -- maybe that helped?